man/tpm2_loadexternal.1.md
% tpm2_loadexternal(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_loadexternal**(1) - Load an external object into the TPM.

# SYNOPSIS

**tpm2_loadexternal** [*OPTIONS*]

# DESCRIPTION

**tpm2_loadexternal**(1) - This command loads an external object into the TPM,
forgoing TPM protections. Ie, the key material is not protected by the parent
object's seed. It can also be used to load TSS2 Private Keys in pem format.
The command allows loading of just the public portion of an object or both the
public and private portions of an object. For TSS2 Private Keys, only the public
portion of the key is loaded.

The tool outputs the name of the loaded object in a YAML dictionary format
with the key *name* where the value for that key is the name of the object
in hex format, for example:
```yaml
name: 000bac25cb8743111c8e1f52f2ee7279d05d3902a18dd1af694db5d1afa7adf1c8b3
```

It also saves a context file for future interactions with the object.

# OPTIONS

  * **-C**, **\--hierarchy**=_OBJECT_:

    Hierarchy to use for the ticket, optional. Defaults to **n**, **null**.
    Supported options are:
      * **o** for the **owner** hierarchy.
      * **p** for the **platform** hierarchy.
      * **e** for the **endorsement** hierarchy.
      * **n** for the **null** hierarchy.

  * **-G**, **\--key-algorithm**=_ALGORITHM_:

    The algorithm used by the key to be imported. Supports:
    * **aes** - AES 128,192 or 256 key.
    * **rsa** - RSA 1024 or 2048 key.
    * **ecc** - ECC NIST P192, P224, P256, P384 or P521 public and private key.

  * **-u**, **\--public**=_FILE_:

    The public portion of the object, this can be one of the following file
    formats:
      * TSS - The TSS/TPM format. For example from option `-u` of command
        **tpm2_create**(1).
      * RSA - OSSL PEM formats. For example `public.pem` from the command
        `openssl rsa -in private.pem -out public.pem -pubout`
      * ECC - OSSL PEM formats. For example `public.pem` from the command
        `openssl ec -in private.ecc.pem -out public.ecc.pem -pubout`

  * **-r**, **\--private**=_FILE_:

    The sensitive portion of the object, optional. If one wishes to use the
    private portion of a key, this must be specified. Like option **-u**, this
    command takes files in the following format:
      * RSA - OSSL PEM formats. For example `private.pem` from the command
        `openssl genrsa -out private.pem 2048`
        Since an RSA public key can be derived from the private PEM file, their
        is no need to specify -u for the public portion.
      * TSS2 PrivateKey PEM formats.

    *Note*: The private portion does not respect TSS formats as it's impossible
    to get a **TPM2B_SENSITIVE** output from a previous command. They are always
    protected by the TPM as **TPM2B_PRIVATE** blobs.

  * **-p**, **\--auth**=_AUTH_:

    The authorization value for the key, optional.

  * **-L**, **\--policy**=_FILE_ or _HEX\_STRING_:

    The input policy file or hex string, optional. A file or hex string
    containing the hash of a policy derived from `tpm2_createpolicy` or
    another policy digest generating source.

  * **-g**, **\--hash-algorithm**=_ALGORITHM_:

    The hash algorithm for generating the objects name. This is optional
    and defaults to sha256 when not specified. However, load external supports
    having a *null* name algorithm. In this case, no cryptographic binding
    checks between the public and private portions are performed.

  * **-a**, **\--attributes**=_ATTRIBUTES_:

    The object attributes, optional. The default for created objects is:
    `TPMA_OBJECT_SIGN_ENCRYPT|TPMA_OBJECT_DECRYPT`. Optionally, if -p is
    specified or no `-p` or `-L` is specified then `TPMA_OBJECT_USERWITHAUTH`
    is added to the default attribute set.

    *Note*: If specifying attributes, the TPM will reject certain attributes
    like **TPMA_OBJECT_FIXEDTPM**, as those guarantees cannot be made.

  * **-c**, **\--key-context**=_FILE_

    The file name to save the object context, required.

  * **-n**, **\--name**=_FILE_:

    An optional file to save the object name, which is in a binary hash format.
    The size of the hash is based on name algorithm or the **-g** option.

  * **\--passin**=_OSSL\_PEM\_FILE\_PASSWORD_

    An optional password for an Open SSL (OSSL) provided input file.
    It mirrors the -passin option of OSSL and is known to support the pass,
    file, env, fd and plain password formats of openssl.
    (see *man(1) openssl*) for more.
    
 * **-e**, **\--rsa_exponent_zero**:
    Set the exponent of a public RSA key to zero, to enable compatibility
    in the computation of the key name if the TPM key is generated with
    exponent zero.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[algorithm specifiers](common/alg.md) details the options for specifying
cryptographic algorithms _ALGORITHM_.

[object attribute specifiers](common/obj-attrs.md) details the options for
specifying the object attributes _ATTRIBUTES_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.


# NOTES

* If the hierarchy is *null* or the name hashing algorithm is *null*, tickets
  produced using the object will be NULL.

* If the private portion of an object is specified, the hierarchy must be *null*
  or the TPM will reject loading it.


# EXAMPLES

## Load a TPM generated public key into the *owner* hierarchy

```bash
tpm2_createprimary -c primary.ctx

tpm2_create -C primary.ctx -u pub.dat -r priv.dat

tpm2_loadexternal -C o -u pub.dat -c pub.ctx
name: 000b9be4d7c6193a57e1bfc86a42a6b03856a91d2f9e77c6cbdb796a783d52d4b3b9
```

## Load an RSA public key into the *owner* hierarchy

```bash
openssl genrsa -out private.pem 2048

openssl rsa -in private.pem -out public.pem -outform PEM -pubout

tpm2_loadexternal -C o -Grsa -u public.pem -c key.ctx
name: 000b7b91d304d16995d42792b57d0fb25df7abe5fdd8afe9950730e00dc5b934ddbc
```

## Load an RSA key-pair into the *null* hierarchy

```bash
openssl genrsa -out private.pem 2048

tpm2_loadexternal -C n -Grsa -r private.pem -c key.ctx
name: 000b635ea220b6c62ec1d02343859dd203c8ac5dad82ebc5b124e407d2502f88691f
```

## Load an AES key into the *null* hierarchy

```bash
dd if=/dev/urandom of=sym.key bs=1 count=16

tpm2_loadexternal -C n -Gaes -r sym.key -c key.ctx
name: 000bfc4d8dd7e4f921bcc9dca4b04f49564243cd9def129a3740002bfd4b9e966d34
```

## Load TSS2 Private Key into the *null* hierarchy

```bash
tpm2_loadexternal -r tss_privkey.pem -c tss_privkey.ctx
name: 000bc5a216702aca9ba226af1214c50dc4dc33ce6269677aa581ea6d9eec7f27000d
```

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/tpm2_loadexternal.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <assert.h>
#include <stdlib.h>
#include <string.h>

#include <openssl/rand.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_alg_util.h"
#include "tpm2_attr_util.h"
#include "tpm2_auth_util.h"
#include "tpm2_hierarchy.h"
#include "tpm2_openssl.h"
#include "tpm2_policy.h"
#include "tpm2_tool.h"
#include "object.h"

#define MAX_SESSIONS 3
typedef struct tpm_loadexternal_ctx tpm_loadexternal_ctx;
struct tpm_loadexternal_ctx {
    /*
     * Inputs
     */
    TPMI_RH_HIERARCHY hierarchy_value;
    ESYS_TR handle;
    char *public_key_path; /* path to the public portion of an object */
    char *private_key_path; /* path to the private portion of an object */
    char *attrs; /* The attributes to use */
    char *auth; /* The password for use of the private portion */
    char *policy; /* a policy for use of the private portion */
    char *name_alg; /* name hashing algorithm */
    char *key_type; /* type of key attempting to load, defaults to an auto attempt */
    char *name_path; /* An optional path to output the loaded objects name information to */
    char *passin; /* an optional auth string for the input key file for OSSL */
    TPM2B_SENSITIVE priv; /* Set the AUTH value for sensitive portion */
    TPM2B_PUBLIC pub; /* Load the users specified public object if specified via -u*/
    bool autoflush; /* Flush the object after creation of the ctx file */
    bool rsa_exponent_zero; /* Set rsa exponent to zero to support name computation for
                               tpm2 keys with exponent zero. */
    /*
     * TSS Privkey related
     */
    bool is_tsspem;

    /*
     * Outputs
     */
    char *context_file_path;
    TPM2B_NAME *name;

    /*
     * Parameter hashes
     */
    const char *cp_hash_path;
    TPM2B_DIGEST cp_hash;
    bool is_command_dispatch;
    TPMI_ALG_HASH parameter_hash_algorithm;
};

static tpm_loadexternal_ctx ctx = {
    /*
     * default to the NULL hierarchy, as the tpm rejects loading a private
     * portion of an object in other hierarchies.
     */
    .hierarchy_value = TPM2_RH_NULL,
    .parameter_hash_algorithm = TPM2_ALG_ERROR,
    .autoflush = false,
};

static tool_rc load_external(ESYS_CONTEXT *ectx) {

    bool is_priv_specified = (ctx.private_key_path != 0 && !ctx.is_tsspem);
    return tpm2_loadexternal(ectx, is_priv_specified ? &ctx.priv : 0, &ctx.pub,
        ctx.hierarchy_value, &ctx.handle, &ctx.cp_hash,
        ctx.parameter_hash_algorithm);
}

static tool_rc process_output(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    /*
     * 1. Outputs that do not require TPM2_CC_<command> dispatch
     */
    bool is_file_op_success = true;
    if (ctx.cp_hash_path) {
        is_file_op_success = files_save_digest(&ctx.cp_hash, ctx.cp_hash_path);

        if (!is_file_op_success) {
            return tool_rc_general_error;
        }
    }

    tool_rc rc = tool_rc_success;
    if (!ctx.is_command_dispatch) {
        return rc;
    }

    /*
     * 2. Outputs generated after TPM2_CC_<command> dispatch
     */
    rc = tpm2_tr_get_name(ectx, ctx.handle, &ctx.name);
    if (rc != tool_rc_success) {
        return rc;
    }
    assert(ctx.name);

    rc = files_save_tpm_context_to_path(ectx, ctx.handle,
         ctx.context_file_path, ctx.autoflush);
    if (rc != tool_rc_success) {
        goto out;
    }

    tpm2_tool_output("name: ");
    tpm2_util_hexdump(ctx.name->name, ctx.name->size);
    tpm2_tool_output("\n");

    if (ctx.name_path) {
        bool result = files_save_bytes_to_file(ctx.name_path, ctx.name->name,
            ctx.name->size);
        if (!result) {
            rc = tool_rc_general_error;
            goto out;
        }
    }

out:
    free(ctx.name);

    return rc;
}

#define BASE_DEFAULT_ATTRS \
    (TPMA_OBJECT_DECRYPT | TPMA_OBJECT_SIGN_ENCRYPT | TPMA_OBJECT_USERWITHAUTH)

static void setup_default_attrs(TPMA_OBJECT *attrs, bool has_policy,
    bool has_auth) {

    /* Handle Default Setup */
    *attrs = BASE_DEFAULT_ATTRS;

    if (has_policy && !has_auth) {
        *attrs &= ~TPMA_OBJECT_USERWITHAUTH;
    }
}

static tool_rc process_inputs(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);
    /*
     * 1. Object and auth initializations
     */

    /*
     * 1.a Add the new-auth values to be set for the object.
     */

    /*
     * 1.b Add object names and their auth sessions
     */

    /*
     * 2. Restore auxiliary sessions
     */

    /*
     * 3. Command specific initializations
     */
    TPMA_OBJECT def_attrs = 0;
    if (!ctx.attrs) {
        setup_default_attrs(&def_attrs, !!ctx.policy, !!ctx.auth);
    }

    tool_rc rc = tool_rc_success;
    if (ctx.is_tsspem) {
        /* Fetch and set public */
        TPM2B_PRIVATE priv = { 0 };
        rc = tpm2_util_object_fetch_priv_pub_from_tpk(ctx.private_key_path,
            &ctx.pub, &priv);
        if (rc != tool_rc_success) {
            /*
             * We only load a TSS format for the public portion, so if
             * someone hands us a public file, we'll assume the TSS format when
             * no -G is specified.
             *
             * If they specify a private they need to tell us the type we expect.
             * This helps reduce auto-guess complexity, as well as future proofing
             * us for being able to load XOR. Ie we don't want to guess XOR or HMAC
             * in leui of AES or vice versa.
             */
            LOG_ERR("Expected key type via -G option when specifying private"
                    " portion of object that is not in tssprivkey format");
            return rc;
        }

        goto priv_path;
    }

    /*
     * Input values are public assumed to be in the TSS format unless:
     * 1. A private (TPM2B_PRIVATE) key is specified. Since a TPM2B_PRIVATE is
     *    nonsense to load externally.
     * 2. The key_type is specified indicated you want to load an external
     *    public raw key of some type
     * 3. The third option is no public, which means the public could be coming
     *    from a private PEM file
     */
    if (!ctx.key_type && ctx.public_key_path) {
        /* Load TSS */
        bool result = files_load_public(ctx.public_key_path, &ctx.pub);
        if (!result) {
            return tool_rc_general_error;
        }

        /* overwrite certain things here */
        if (ctx.name_alg) {
            ctx.pub.publicArea.nameAlg = tpm2_alg_util_from_optarg(ctx.name_alg,
                tpm2_alg_util_flags_hash);
            if (ctx.pub.publicArea.nameAlg == TPM2_ALG_ERROR) {
                LOG_ERR("Invalid name hashing algorithm, got\"%s\"",
                    ctx.name_alg);
                return tool_rc_unsupported;
            }
        }

        if (ctx.attrs) {
            bool res = tpm2_attr_util_obj_from_optarg(ctx.attrs,
                    &ctx.pub.publicArea.objectAttributes);
            if (!res) {
                return tool_rc_unsupported;
            }
        }

        if (ctx.policy) {
            rc = tpm2_policy_set_digest(ctx.policy,
                    &ctx.pub.publicArea.authPolicy);
            if (rc != tool_rc_success) {
                return rc;
            }
        }


    } else if (ctx.public_key_path || ctx.key_type) {
        /* Load RAW public */
        TPM2B_PUBLIC template = { 0 };
        rc = tpm2_alg_util_public_init(ctx.key_type, ctx.name_alg,
            ctx.attrs, ctx.policy, def_attrs, &template);
        if (rc != tool_rc_success) {
            return rc;
        }

        ctx.pub = template;

        if (ctx.public_key_path) {
            /* Get the public from a RAW source, ie PEM */
            bool result = tpm2_openssl_load_public(ctx.public_key_path,
                template.publicArea.type, &ctx.pub);
            if (!result) {
                return tool_rc_general_error;
            }
            if (ctx.pub.publicArea.type == TPM2_ALG_RSA && ctx.rsa_exponent_zero) {
                ctx.pub.publicArea.parameters.rsaDetail.exponent = 0;
            } else if (ctx.rsa_exponent_zero) {
                LOG_WARN("Option --rsa_exponent_zero used for key which is no RSA key.");
            }
        }
    } else {
        LOG_ERR("Unkown internal state");
        return tool_rc_general_error;
    }

priv_path:
    /*
     * Okay, we have the public portion in some form, at a minimum a template
     * and at a maximum a fully specified public, load the private portion.
     */
    if (ctx.private_key_path && !ctx.is_tsspem) {
        /*
         * when nameAlg is not TPM2_ALG_NULL, seed value is needed to pass
         * consistency checks by TPM
         */
        TPM2B_DIGEST *seed = &ctx.priv.sensitiveArea.seedValue;
        seed->size = tpm2_alg_util_get_hash_size(ctx.pub.publicArea.nameAlg);
        if (seed->size != 0) {
            RAND_bytes(seed->buffer, seed->size);
        }

        tpm2_openssl_load_rc load_status = tpm2_openssl_load_private(
            ctx.private_key_path, ctx.passin, ctx.auth, &ctx.pub, &ctx.pub,
            &ctx.priv);
        if (load_status == lprc_error) {
            return tool_rc_general_error;
        }

        /*
         * If we cannot load the public from the private and a path
         * is not specified for public, this is an error.
         *
         * If we loaded the public from the private and a public was
         * specified, this is warning. re-init public and load the
         * specified one.
         */
        bool is_ossl_pub_load = tpm2_openssl_did_load_public(load_status);
        if (!is_ossl_pub_load && !ctx.public_key_path) {
            LOG_ERR("Only loaded a private key, expected public key in either"
                    " private PEM or -r option");
            return tool_rc_general_error;
        }

        if (is_ossl_pub_load && ctx.public_key_path) {
            /*
             * Okay... we got two publics, use the user specified one and
             * issue a warning
             */
            LOG_WARN("Loaded a public key from the private portion"
                    " and a public portion was specified via -u. Defaulting"
                    " to specified public");

            /*
             * This effectively turns off cryptographic binding between
             * public and private portions of the object.
             */
            ctx.pub.publicArea.nameAlg = TPM2_ALG_NULL;
        }
    }

    /*
     * 4. Configuration for calculating the pHash
     */

    /*
     * 4.a Determine pHash length and alg
     */
    tpm2_session *all_sessions[MAX_SESSIONS] = {
        0,
        0,
        0
    };

    const char **cphash_path = ctx.cp_hash_path ? &ctx.cp_hash_path : 0;

    ctx.parameter_hash_algorithm = tpm2_util_calculate_phash_algorithm(ectx,
        cphash_path, &ctx.cp_hash, 0, 0, all_sessions);
    /*
     * 4.b Determine if TPM2_CC_<command> is to be dispatched
     */
    ctx.is_command_dispatch = ctx.cp_hash_path ? false : true;

    return rc;
}

static tool_rc check_options(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    if (!ctx.public_key_path && !ctx.private_key_path) {
        LOG_ERR("Expected either -r or -u options");
        return tool_rc_option_error;
    }

    if (!ctx.context_file_path && !ctx.cp_hash_path) {
        LOG_ERR("Expected -c option");
        return tool_rc_option_error;
    }

    ctx.is_tsspem = (!ctx.key_type && ctx.private_key_path);

    return tool_rc_success;
}

static bool on_option(char key, char *value) {

    bool result;

    switch (key) {
    case 'C':
        result = tpm2_util_handle_from_optarg(value, &ctx.hierarchy_value,
                TPM2_HANDLE_FLAGS_ALL_HIERACHIES);
        if (!result) {
            return false;
        }
        break;
    case 'u':
        ctx.public_key_path = value;
        break;
    case 'r':
        ctx.private_key_path = value;
        break;
    case 'c':
        ctx.context_file_path = value;
        break;
    case 'a':
        ctx.attrs = value;
        break;
    case 'p':
        ctx.auth = value;
        break;
    case 'L':
        ctx.policy = value;
        break;
    case 'g':
        ctx.name_alg = value;
        break;
    case 'G':
        ctx.key_type = value;
        break;
    case 'n':
        ctx.name_path = value;
        break;
    case 0:
        ctx.passin = value;
        break;
    case 1:
        ctx.cp_hash_path = value;
        break;
    case 'R':
        ctx.autoflush = true;
        break;
    case 'e':
        ctx.rsa_exponent_zero = true;
        break;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {

    const struct option topts[] = {
      { "hierarchy",      required_argument, 0, 'C'},
      { "public",         required_argument, 0, 'u'},
      { "private",        required_argument, 0, 'r'},
      { "key-context",    required_argument, 0, 'c'},
      { "attributes",     required_argument, 0, 'a'},
      { "policy",         required_argument, 0, 'L'},
      { "auth",           required_argument, 0, 'p'},
      { "hash-algorithm", required_argument, 0, 'g'},
      { "key-algorithm",  required_argument, 0, 'G'},
      { "name",           required_argument, 0, 'n'},
      { "passin",         required_argument, 0,  0 },
      { "cphash",         required_argument, 0,  1 },
      { "autoflush",      no_argument,       0, 'R'},
      { "rsa_exponent_zero", no_argument,    0, 'e'},
    };

    *opts = tpm2_options_new("C:u:r:c:a:p:L:g:G:n:Re", ARRAY_LEN(topts), topts,
        on_option, 0, 0);

    return *opts != 0;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {

    UNUSED(flags);

    /*
     * 1. Process options
     */
    tool_rc rc = check_options(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 2. Process inputs
     */
    rc = process_inputs(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 3. TPM2_CC_<command> call
     */
    rc = load_external(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    /*
     * 4. Process outputs
     */
    return process_output(ectx);
}

static tool_rc tpm2_tool_onstop(ESYS_CONTEXT *ectx) {

    UNUSED(ectx);

    /*
     * 1. Free objects
     */

    /*
     * 2. Close authorization sessions
     */

    /*
     * 3. Close auxiliary sessions
     */

    return tool_rc_success;
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("loadexternal", tpm2_tool_onstart, tpm2_tool_onrun,
    tpm2_tool_onstop, 0)
